middleware.ts

```diff
@@ -2,28 +2,67 @@ import { getIronSession } from 'iron-session';
 import type { NextRequest } from 'next/server';
 import { NextResponse } from 'next/server';
 import { ISession } from '#models/user/session';
-import { extractSirenOrSiretSlugFromUrl } from '#utils/helpers';
+import {
+  extractSirenOrSiretSlugFromUrl,
+  isLikelyASiretOrSiren,
+} from '#utils/helpers';
 import { sessionOptions, setVisitTimestamp } from '#utils/session';
 
-// This function can be marked `async` if using `await` inside
-export async function middleware(request: NextRequest) {
-  const pathname = request.nextUrl.pathname;
-  const sirenOrSiretSlug = extractSirenOrSiretSlugFromUrl(pathname);
+const shouldRedirect = (path: string, search: string, url: string) => {
+  try {
+    const sirenOrSiretSlug = extractSirenOrSiretSlugFromUrl(path);
+    if (sirenOrSiretSlug) {
+      if (path.startsWith('/entreprise/')) {
+        if (!isLikelyASiretOrSiren(sirenOrSiretSlug)) {
+          return new URL(`/404`, url);
+        }
 
-  const isEntreprisePage = pathname.startsWith('/entreprise/');
+        if (sirenOrSiretSlug.length === 14) {
+          return new URL(`/etablissement/${sirenOrSiretSlug}`, url);
+        }
+      }
 
-  if (isEntreprisePage && sirenOrSiretSlug.length === 14) {
-    return NextResponse.redirect(
-      new URL(`/etablissement/${sirenOrSiretSlug}`, request.url)
-    );
+      if (path.startsWith('/etablissement/')) {
+        if (!isLikelyASiretOrSiren(sirenOrSiretSlug)) {
+          return new URL(`/404`, url);
+        }
+        if (sirenOrSiretSlug.length === 9) {
+          return new URL(`/entreprise/${sirenOrSiretSlug}`, url);
+        }
+      }
+    }
+
+    if (path.startsWith('/rechercher')) {
+      const sirenOrSiretParam = extractSirenOrSiretSlugFromUrl(
+        search.replaceAll(/[+]|(%20)/g, '')
+      );
+
+      if (isLikelyASiretOrSiren(sirenOrSiretParam)) {
+        if (sirenOrSiretParam.length === 14) {
```

Comment:
Maybe a `isLikelyASiren` and `isLikelyASiret` would be a nice abstraction ? (not 100% sure)